posts/change-data-capture-cloudsql-postgres-with-apache-airflow-1.md

```diff
@@ -0,0 +1,1002 @@
+
+<span style="display:block;text-align:center;">
+  <img src="../assets/cdc-cloudsql-1/cdc_cloudsql_airflow_2021-01-19_at_2.48.37_PM.png" width=400> 
+</span>
+
+The `start_date` you set in the DAG definition follows this logic. If you choose:
+
+```python
+start_date = DateTime(2021, 1, 14, 13, 0, 0),
+schedule_interval = '@hourly',
+catchup = False
+```
+
+you would see the first run at `2021-01-14T14:00:00+00:00` if the current date were before that and before the `start_date`. If the current calendar date was after the `start_date`, you would see the first run start at the beginning of the most recent hour. 
+
+> This is powerful because you can use that hard interval to further refine your watermarks and obtain data created **right up until the DAG starts running**. With that, you now have near-real time change data capture capabilities. Furthermore, the consistency and high-availability allows for SLAs to be set on this data (and in Airflow), something extremely difficult in a streaming solution.
+
+To do this:
+
+1. If your interval is always going to be the same, in `table_1.sql`, you could hardcode your `low_watermark` adjustment to be:
+    ```python
+    {%- set low_watermark =  execution_date -%}
+    {%- set high_watermark =  execution_date.add(hours=1) -%}
+    SELECT * FROM some_schema.table_1
+    WHERE event_timestamp AT TIME ZONE 'UTC' >= '{{ low_watermark }}' AT TIME ZONE 'UTC'
+    AND event_timestamp AT TIME ZONE 'UTC' < '{{ high_watermark }}'
+    ```
+
+    which you would see renders to
+
+    ```bash
+    # ----------------------------------------------------------
+    # property: export_sql
+    # ----------------------------------------------------------
+    SELECT * FROM some_schema.table_1
+    WHERE event_timestamp AT TIME ZONE 'UTC' >= '2021-01-14T13:00:00+00:00' AT TIME ZONE 'UTC'
+    AND event_timestamp AT TIME ZONE 'UTC' < '2021-01-14T14:00:00+00:00'
+    ```
+
+    getting you changes in the previous hour. 
+
+2. Make the following changes to the rest of your `.sql` files.
+
+3. In your DAG definition, adjust to your interval of choice. Suppose you want to make this run every 30 minutes. The `schedule_interval` parameter also accepts a crontab value. So in the DAG definition, that would be `schedule_interval = '* /30 * * * *'`, and your templated SQL would appear:
+
+    ```bash
+    {%- set low_watermark =  execution_date -%}
+    {%- set high_watermark =  execution_date.add(minutes=30) -%}
+    SELECT * FROM some_schema.table_1
+    WHERE event_timestamp AT TIME ZONE 'UTC' >= '{{ low_watermark }}' AT TIME ZONE 'UTC'
+    AND event_timestamp AT TIME ZONE 'UTC' < '{{ high_watermark }}'
+    ```
+
+    In a future post, I will show you how to set these dynamically, through the UI, as to allow interval changes and full scans when needed.
```

Comment:
Don't think we need the commas here around `through the UI`.

"In a future post, I will show you how to set these dynamically through the UI to allow interval changes and full scans."